### PR TITLE
Non-web health checks

### PR DESCRIPTION
**Pull request by pschoenfelder**

### CODE REVIEW

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorMonitor.java [hunk 1]
```diff
@@ -83,6 +102,10 @@ public SingularityExecutorMonitor(@Named(SingularityExecutorModule.ALREADY_SHUT_
     this.threadChecker = threadChecker;
     this.threadChecker.start(this);
 
+    this.requestManager = requestManager;
+    this.deployManager = deployManager;
+    this.taskManager = taskManager;
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorMonitor.java]: SingularityData stuff would be a very heavy dep to add into every executor. I do not think we want to add these here.
- `requestManager` data your fetching here isn't actually being used
- `HealthcheckOptions` we should feed in through the `SingularityTaskExecutorData` like you are doing with the path
- There is no valuable information in saving health check results for the task since all we are doing is checking a file. There will essentially be a single pass/fail health check result, and we can even leave that as encoded in the task status update to keep things simpler if we want. The basics would be that the status update of running serves as our healthcheck success, or a status update of killing/failed/etc serves as our failure.

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorModule.java [hunk 2]
```diff
@@ -38,6 +39,7 @@
 
   @Override
   protected void configure() {
+    install(new SingularityServiceModule());
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorModule.java]: side note to bringing in this dep, installing this module would also likely start up a number of pollers/db connections/zk connections/etc which we don't want

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorMonitor.java [hunk 3]
```diff
@@ -328,6 +351,52 @@ private void onSuccessThrows(ProcessBuilder processBuilder) {
...
+                    healthCheckEnd,
+                    Optional.absent(),
+                    Optional.of("Task timed out on health checks."),
+                    taskId,
+                    Optional.of(true)));
+                return;
+              }
+            }
+
             processRunningTasks.put(task.getTaskId(), submitProcessMonitor(task, processBuilder));
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorMonitor.java]: The task process doesn't actually start running until after this line. We'd want to make sure the process + monitors are up and running before we start looking at health checks. Ideally it is after we start and right before the TASK_RUNNING. Something like [here](https://github.com/HubSpot/Singularity/blob/master/SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java#L27) I think would make it much simpler.

File: SingularityBase/src/main/java/com/hubspot/singularity/SingularityDeploy.java [hunk 4]
```diff
@@ -89,6 +89,7 @@
   private final Optional<Long> healthcheckMaxTotalTimeoutSeconds;
 
   private final Optional<HealthcheckOptions> healthcheck;
+  private final Optional<String> healthCheckResultFilePath;
```
Comment (ssalinas) [inline SingularityBase/src/main/java/com/hubspot/singularity/SingularityDeploy.java]: I think we can make this part of the `HealthcheckOptions` object. There are healthcheck-related fields outside that object, but in general they are deprecated, in favor of the ones consolidated in `HealthcheckOptions`

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java [hunk 5]
```diff
@@ -226,6 +226,9 @@
   @JsonProperty
   private boolean useFileAttributes = false;
 
+  @JsonProperty
+  private int healthcheckMaxRetries = 30;
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java]: Maybe name this `default...` instead? Makes it clearer as to why there is a setting here

File: SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/config/SingularityExecutorCleanupConfiguration.java [hunk 6]
```diff
@@ -71,6 +72,12 @@
   @NotNull
   private String defaultS3Bucket = "";
 
+  @NotNull
+  private Optional<String> healthCheckResultFilePath = Optional.absent();
```
Comment (ssalinas) [inline SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/config/SingularityExecutorCleanupConfiguration.java]: why are these needed in executor cleanup? It should have no interaction with health checks

File: SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/SingularityExecutorCleanup.java [hunk 7]
```diff
@@ -215,7 +215,9 @@ private SingularityExecutorTaskDefinition withDefaults(SingularityExecutorTaskDe
             oldDefinition.getExecutorData().getRequestGroup(),
             oldDefinition.getExecutorData().getS3StorageClass(),
             oldDefinition.getExecutorData().getApplyS3StorageClassAfterBytes(),
-            oldDefinition.getExecutorData().getCpuHardLimit()
+            oldDefinition.getExecutorData().getCpuHardLimit(),
+            oldDefinition.getExecutorData().getHealthCheckResultFilePath().isPresent() ? oldDefinition.getExecutorData().getHealthCheckResultFilePath() : cleanupConfiguration.getHealthCheckResultFilePath(),
```
Comment (ssalinas) [inline SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/SingularityExecutorCleanup.java]: cleanup shouldn't need to interact with the check at all, can just set it to either `oldDefinition.getExecutorData().get....` or just to absent

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java [hunk 8]
```diff
@@ -39,4 +58,27 @@ public String toString() {
...
+      try {
+        Integer healthcheckMaxRetries = maybeOptions.get().getMaxRetries().or(configuration.getHealthcheckMaxRetries());
+
+        Retryer<String> retryer = RetryerBuilder.<String>newBuilder()
+            .retryIfResult(path -> !new File(path).exists())
+            .withWaitStrategy(WaitStrategies.fixedWait(1L, TimeUnit.SECONDS))
+            .withStopStrategy(StopStrategies.stopAfterAttempt(healthcheckMaxRetries))
+            .build();
+
+        retryer.call(() -> expectedHealthCheckResultFilePath.get());
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java]: would it make more sense for the exists to be here, and the retryer to a have a retryIf predicate instead (i.e. retry if false)?

File: SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/SingularityExecutorCleanup.java (hunk 7 above)
Comment (pschoenfelder) [reply SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/SingularityExecutorCleanup.java]: This constructs a new `SingularityExecutorTaskDefinition` and I was following the pattern here. I would think just sending `Optional.absent()` would be okay too though. Thougths?

Comment (ssalinas) [reply SingularityExecutorCleanup/src/main/java/com/hubspot/singularity/executor/cleanup/SingularityExecutorCleanup.java]: absent would be fine here since it should be unused in this context

**Comment by pschoenfelder** [comment]
@ssalinas What do you think the best way to test this is?

**Comment by ssalinas** [comment]
Will likely just need to create a sample worker in our staging cluster. Can make the command something easy like `sleep 30 && touch ./healthy && sleep 1000` to give you time to look at what's going on

**Comment by pschoenfelder** [comment]
Finally got this working (thanks for the extra 👀@ssalinas).

One last bit I want to resolve is the difference between directories. The directory where the command is run is different from where the health check takes place. What directory do users typically expect to use? If I were a user I wouldn't want to worry about the directory structure for this, and would assume these run from a common directory.
```
    "executorData":{
        "cmd":"bash -c 'sleep 5 && touch ../../healthy.out'",
    },
    "healthcheck": {
        "healthcheckResultFilePath": "healthy.out"
    }
```

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java [hunk 9]
```diff
@@ -39,4 +61,32 @@ public String toString() {
     return "SingularityExecutorTaskProcessCallable [task=" + task + "]";
   }
 
+  private void runHealthcheck() {
+    Optional<HealthcheckOptions> maybeOptions = task.getTaskDefinition().getHealthcheckOptions();
+    Optional<String> expectedHealthcheckResultFilePath = task.getTaskDefinition().getHealthcheckResultFilePath();
+
+    String taskAppDirectory = task.getTaskDefinition().getTaskAppDirectory();
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java]: nit, doesn't seem like this needs to be outside the if statement

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java [hunk 10]
```diff
@@ -39,4 +61,31 @@ public String toString() {
...
+      LOG.debug("Checking for healthcheck file {}", expectedHealthcheckResultFilePath.get());
+      String taskAppDirectory = task.getTaskDefinition().getTaskAppDirectory();
+      File fullHealthcheckPath = Paths.get(taskAppDirectory, expectedHealthcheckResultFilePath.get()).toFile();
+
+      try {
+        Integer healthcheckMaxRetries = maybeOptions.get().getMaxRetries().or(configuration.getDefaultHealthcheckMaxRetries());
+
+        Retryer<Boolean> retryer = RetryerBuilder.<Boolean>newBuilder()
+            .retryIfResult(bool -> !bool)
+            .withWaitStrategy(WaitStrategies.fixedWait(1L, TimeUnit.SECONDS))
```
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java]: missed this one earlier. There is a healthcheck interval in the healthcheck options as well. Right now we are essentially treating the max retries as a timeout in seconds, which might be . bit confusing to the user given how the web health checks work. If we take the interval into account here then the math for total timeout works the same as it does for web service checks

**Comment by ssalinas** [comment]
🚢
